designs/node-upgrades.md

```diff
@@ -0,0 +1,185 @@
+
+## How will we upgrade nodes?
+
+When upgrading a node, Karpenter will minimize the downtime of the applications on the node by first creating the replacement node before terminating the old node. Once the replacement is ready, Karpenter will cordon and drain the old node, terminating it when it’s fully drained. Once the old node has been terminated, the upgrade will be complete.
+
+Upgrading nodes without a dedicated controller to orchestrate the logic means relying on the termination controller to scale down and the provisioning controller to scale back up. Since the termination controller is asynchronous and terminates in quick succession with no PDBs. upgrades would happen in parallel, with rate-limiting happening at the pod level. In creating a dedicated controller to orchestrate the logic, there are two options to upgrading nodes - serially and in parallel.
+
+### Option 1 - Upgrade Serially
+
+Karpenter will only upgrade one node at a time. Each node's upgrade will be complete when a its replacement is ready and the old node is terminated. This naturally rate-limits node upgrades to node initialization time and means that users will only have to worry about one node disruption at a time with upgrades.
+
+With serial upgrades, new nodes that cannot become ready or old nodes that cannot fully drain can stop the upgrade process altogether. For example, an old node could fail to fully drain due to a PDB or a do-not-evict annotation, which may require manual intervention. A new node could fail to become ready for many reasons. For one, if the new AMI is incompatible with the cluster or applications, Karpenter should not upgrade more nodes as the more upgrades to that AMI would likely fail. Yet on the other hand, sometimes a failed AMI can succeed in other cases like with an [Outdated CNI](https://karpenter.sh/v0.16.3/troubleshooting/#nodes-stuck-in-pending-and-not-running-the-kubelet-due-to-outdated-cni) - someone using a newer instance type could experience an issue where an older instance type would succeed.
+
+In order to not permanently block serial upgrades, upgrades will have timeouts. Relying on Kubernetes native concepts, Karpenter will compute a drain timeout that respects the GracefulTerminationPeriodSeconds on the pods on the nodes by summing all GracefulTerminationPeriodSeconds fields. This represents the worst case where Karpenter can only evict one pod at a time, taking as long as possible. If the replacement node fails to create, we can terminate the failed upgraded node and try another one, where rolling back is already complete.
+
+### Option 2 - Upgrade in Parallel
+
+For some use cases, upgrading serially may be too slow. To combat this, Karpenter can upgrade in parallel, allowing more than one node to upgrade at a time. In contrast to upgrading serially,  upgrading in parallel requires taking an opinion on hyper-parameters or allowing an API for users to configure them.
+
+#### Option 2a - In parallel with controls
```

Comment:
This feels like the "correct" approach and native to what users expect when it comes to K8s resource management